include/watchdog.h

```diff
@@ -1,116 +1,206 @@
 /*
- * Copyright (c) 2015 Intel Corporation.
+ * Copyright (c) 2017 Nordic Semiconductor ASA
  *
  * SPDX-License-Identifier: Apache-2.0
  */
 
+/**
+ * @file
+ * @brief Public API for watchdog drivers.
+ */
+
+#if defined(CONFIG_WDT_LEGACY_API)
+
+/*
+ * This is the default, and will be the way until the new API below
+ * will be enforced everywhere.
```

Comment:
What exactly is the plan for this?
We have been carrying around "spi" and "spi_legacy" for quite some time, and the legacy API is *still* the default..

Reply:
@andrewboie Well, I guess the reason is that no one feels responsible for updating the drivers.

Reply:
> @andrewboie Well, I guess the reason is that no one feels responsible for updating the drivers.

When changing  API changes we can't expect drivers to be changed over-night to an API that is still in review.  If we just change the APIs to support new IPs and h without looking at the existing drivers we might end up with drivers that can't be moved to the new API easily, looking at existing drivers should be part of the change, or there should be at least a plan.

For starters, we need to have a list of issues added for each driver that needs to be changed, preferably with some instructions for driver owner explaining what changed and what to look for when moving to the new API.

Do we really have to create this _legacy header? Can't we change the API inline deprecating the old APIs? I know it was done for SPI, does not mean it is needed for every API change we do? Maybe it is easier with _legacy.h when implementing the API, but adds lots of complexity and confusion and old API are kept alive and are not deprecated proper.

Reply:
> Can't we change the API inline deprecating the old APIs?

@nashif 
You mean making one watchdog.h file and adding deprecated suffix/prefix to old stuff?

Reply:
yes, that has been the right way to change APIs. Not sure if that is possible at all, but at least it gives us an easier migration path without introducing the LEGACY craft.

Reply:
Isn't that __deprecated was making the built failing? At least it used to be like that when we were still using gerrit (the generated warning was the culprit I mean).

The other problem comes also in naming: deprecating old API will still make old names impossible to be reused it new API. That's not nice: so wdt_reload will be deprecated for instance, and what name to use for the new one?

Reply:
__deprecated should not fail the build, this was changed some time ago.

re naming, yes, that will be tricky. Right now however moving everything to be legacy, i.e. renaming driver names to _legacy and declaring this as the new API without having a single driver use it is not helping. Without actually looking at the existing drivers and providing example of how to migrate and working with the driver owners to move to the new API this will be a problem. 
We do not want to end up with an API that favours one HW and ignores the others, and that was the case we had initially given that the API was put together to address the first WD driver we introduced to Zephyr.

Reply:
I agree, a driver implementing this new API in the same PR would be nice.
Renaming existing drivers with  _legacy is not needed, only the dependency in Kconfig is.

In case of spi, responding to @andrewboie, the legacy is the default because of qmsi and the fact that the related hw was our reference platform. I guess, knowing these hw fate now, we can make the new API the default. That said, even new platform like nrf5 is still made on top of old API.
We need to push driver maintainers to go with new APIs asap.

Reply:
not as good as wdt_reload() but well, if we need a new, let's have a new name.

Reply:
wdt_feed() is better. We do not want to get sued by defenders of animal rights.